### ISSUE TITLE
Predefined breaks for LinearColorMapper

### ISSUE THREAD

**Issue by lrdsouza** [bokeh/bokeh#10656]
I'm developing an application to analyze arbovirus infections in the neighborhoods of two cities. Heat maps are generated for each city, their color scheme varies according to the count of infections in the neighborhoods. One of the problems in generating color of these maps is the interest data distribution. The vector of infection counts for each city has very distant values, such as the vector `v1 = [1, 2, 3, 3, 4, 7, 8, 9, 15, 20, 39, 40]`. The `LinearColorMapper` with 4 colors would generate mapping intervals with size 10 breaks, such as `[1, 10), ..., [30, 40]`. So, as we can see, more than half of `v1` would be included in the break `[1, 10)`, causing many neighborhoods to be mapped to a single color. An ideal set of intervals for this problem would be `c = {[1, 4], [7, 9], [15, 20], [39, 40]}`, as it would allow for more color abundance on maps. An alternative to solve this problem would be allow the `LinearColorMapper` function to receive externally generated intervals or to add a resource in the `LinearColorMapper` function that allows selecting the algorithm for generating breaks, such as Fisher-Jenks and Jenks-Caspall.

**Comment by jbednar** [comment]
The [EqHistColorMapper](https://docs.bokeh.org/en/latest/docs/reference/models/mappers.html#bokeh.models.mappers.EqHistColorMapper) is one approach to solving this issue, balancing the color range across the set of values automatically.

**Comment by bryevdv** [comment]
@jbednar right but it defines the bins for you, correct? I think the ask is for something one level lower (to implement something *like* `EqHistMapper`, but different)

**Comment by jbednar** [comment]
Right; EqHistColorMapper only addresses the last bit: "add a resource ... that allows selecting the algorithm for generating breaks". It doesn't use a piecewise-linear approach as outlined by the OP, and doesn't allow the user to specify anything but the start and end of the range, if that level of control is required, but it does ensure that the colors are mapped well in that range.

**Comment by lrdsouza** [comment]
That's right. I think that adding a parameter to the `LinearColorMapper` function that allows you to select different algorithms for generating these breaks can be of great help. It would be something similar to the `scheme` parameter of the GeoPandas `plot` function.

**Comment by bryevdv** [comment]
>  I think that adding a parameter to the LinearColorMapper function that allows you to select different algorithms for generating these breaks can be of great help

So I am personally 👎 on this sort of approach. It is limiting in the sense that the only algorithms that would be available would be the ones we happen to add, and we would also be on the hook for maintaining the. I would propose something more generic that simply exposes the ability for users to specify piecewise linear maps explicitly. Then generating those maps, using whatever algorithm desired, is the user's responsibility.

**Comment by lrdsouza** [comment]
Yes, really. Thinking in this way, the task of selecting and implementing several algorithms in `LinearColorMapper` is exaggerated, considering that probably the current mappings are sufficient for most Bokeh users. I think a solution could be: `LinearColorMapper` generates equal breaks by default and, if necessary, the user could manually pass a list/array of mappings obtained externally.

**Comment by bryevdv** [comment]
I'm not sure it would make sense to bolt this on to the existing `LinearColorMapper` I think it would have to be a new `SegmentedColorMapper` or similar. Then questions:

* Does each bin only get one color (as described in OP)? or more generally
* Does each bin get an entire palette to linearly interpolate in?

The more general approach would make arbitrary diverging / segmented colormaps more feasible. 

Here's a very rough sketch:

```python
class ColorMapSegment:

   start = Number
   end = Number
   palette = List[Color]

class SegmentedColorMapper:
    segments = List[ColorMapSegment]
    high_color = Color # above the highest segment end
    low_color = Color # below the lowest segment start
    nan_color = Color # for nan values
  
    # for any values not in any segment or covered by high/low/nan
    default_color = Color 
```

I would say it is up to users to ensure segments cover an entire range without gaps (if that is what is desired, otherwise `default_color`  can fill in between gaps) and also that ensuring that segments do not overlap. 

cc @jbednar @bokeh/dev for thoughts

**Comment by lrdsouza** [comment]
For my problem, each bin having a different color is enough. However, i believe that it is more recommended that each bin has a palette for interpolation in problems that require greater granularity.

**Comment by jbednar** [comment]
If you just want a different color per bin, I would assume EqHistColorMapper would already work for your approach. Try it!  

I agree that functionality like this wouldn't belong in LinearColorMapper, which doesn't need that complexity, and that SegmentedColorMapper seems like a better place for it. But I'm a bit confused how piecewise linear maps could work; in practice isn't it necessary to recompute those pieces dynamically (in JS) as the user zooms in and out of a dataset (and thus covers different ranges)? Seems like one would need to implement the underlying break-generating algorithms in JS for it to fit in nicely.

**Comment by philippjfr** [comment]
@jbednar Not following why the intervals would have to be computed dynamically, could you explain that? I can see that being useful in certain cases but I'm not following why fixed intervals wouldn't be useful in itself.

**Comment by jbednar** [comment]
The OP suggested "algorithm for generating breaks, such as Fisher-Jenks and Jenks-Caspall", which to me implies a data-driven approach where the breaks are calculated from the data being shown. Given that the data being shown depends on the viewport, if the breaks are not dynamically computed then they won't reflect the data being shown in that viewport. Sure, a fixed set of breaks is better than not being able to specify the breaks, but it doesn't seem like a full solution to what is being asked for.

**Comment by philippjfr** [comment]
Oh right yes, if we want to support all of those approaches in one color mapper it would have to accept explicit intervals and the ability to choose an algorithm to compute them dynamically but even then you should be able to compute it once across the whole dataset (the default) or dynamically while zooming. In either case the implementation of the algorithms should be JS-side.

We sort of fake the explicit intervals case in HoloViews using this function currently: https://github.com/holoviz/holoviews/blob/a2b37dc70228e34ffb6fe329f0854b1ad6502636/holoviews/plotting/util.py#L919 but for highly divergent data this doesn't work well since we limit the palette to 255 colors.

**Comment by bryevdv** [comment]
Just to be clear I think viewport-based colormapping is out of scope. I think it is far more common for people to want to apply a colormap globally.

**Comment by itsgifnotjiff** [comment]
Hmmm is there an example I can follow to show a `hvplot.quadmesh` with a list of 10 colors corresponding to the 10 ranges created by 10 thresholds? (t1-t2: color1, t2-t3: color2, ..., t9-t10:color9, t10-max color10) @jbednar @philippjfr @ahuang11

![image](https://github.com/bokeh/bokeh/assets/50274281/6c7c0787-7424-4460-8b5f-3a5f45ce37ed)

**Comment by ahuang11** [comment]
Something like this?
https://twitter.com/at_dot_Py/status/1660853256229101569
<img width="170" alt="image" src="https://github.com/bokeh/bokeh/assets/15331990/1bc8a319-f3d5-4bed-b376-77199bb848f9">
<img width="371" alt="image" src="https://github.com/bokeh/bokeh/assets/15331990/2f475b5f-79c1-4874-8944-d08369a79e2a">

Or more complex:
<img width="262" alt="image" src="https://github.com/bokeh/bokeh/assets/15331990/83e95ab8-8828-48f8-b62a-9760ac7a530c">
For
<img width="33" alt="image" src="https://github.com/bokeh/bokeh/assets/15331990/b538d158-4536-492b-bad4-bf52d3dc2ef6">

**Comment by ahuang11** [comment]
I think this is what you're looking for:
1. uneven tick intervals
2. select color mapping
3. tick mark locations
```
import xarray as xr
import hvplot.xarray
from bokeh.models import FixedTicker

ds = xr.tutorial.open_dataset('air_temperature')

ds.hvplot("lon", "lat").opts(
    cmap=["green", "yellow", "orange", "red"],
    colorbar_opts={"ticker": FixedTicker(ticks=[230, 250, 270, 300, 320])},
    color_levels=[230, 250, 270, 300, 320],
    clim=(230, 320),
)
```
<img width="1050" alt="image" src="https://github.com/bokeh/bokeh/assets/15331990/1e715fbf-54f1-4604-bf96-1232f95df47f">

Implementation of `cticks` in the future might make this easier as well:
https://github.com/holoviz/holoviews/issues/5723

**Comment by itsgifnotjiff** [comment]
It seems you are putting me on the right track Mr. @ahuang11 but not quite there maybe because of `hvplot`. I use a quadmesh as my dataset is an Xarray Dataset and I need to project it and rasterize because of the size. What am I missing to mirror your example?

```python
colors = ['#cccccc', '#999999', '#7489a8','#00ccff','#0000fe','#ffff00','#dbc903','#fa8484','#990001','#000000']
thresholds = [2, 5, 10, 20, 40, 60, 80, 100, 200, 500]
```

I think the problem comes from how I define the `cticks` and `clim` I need to define them as in the graph above where the last color (black) colors anything above 500. It is very computationally expensive to find the max of an Xarray Dataset so i was wondering if there is another way Mr. @ahuang11 ?


![image](https://github.com/bokeh/bokeh/assets/50274281/adb2ec9e-116f-40dd-ae65-1b7c8bfd6c93)

**Comment by ahuang11** [comment]
I think you're mixing up putting kwargs in the hvplot call vs `opts`

Once you migrate some of the keywords from hvplot to opts, it should work, like this:
```
import xarray as xr
import hvplot.xarray
from bokeh.models import FixedTicker

ds = xr.tutorial.open_dataset('air_temperature')

ds.hvplot.quadmesh("lon", "lat").opts(
    cmap=["green", "yellow", "orange", "red"],
    colorbar_opts={"ticker": FixedTicker(ticks=[230, 250, 270, 300, 320])},
    color_levels=[230, 250, 270, 300, 320],
    clim=(230, 320),
)
```

<img width="851" alt="image" src="https://github.com/bokeh/bokeh/assets/15331990/a15e899d-df93-4e36-81a7-25139e2bf8df">

**Comment by bryevdv** [comment]
Hi @ahuang11 @itsgifnotjiff this seems to have veered more towards a support discussion, please migrate to https://discourse.bokeh.org for Bokeh-specific discussions or https://discourse.holoviz.org for things like hvplot.

**Comment by itsgifnotjiff** [comment]
@bryevdv I am sorry I have a very particular case and used this to understand how the `opts` affect the underlying `Holoviews` objects. Exposing the `cticks` and other relevant properties seems like a necessary step to make `hvplot` more useful for geospatial scientists as they are very particular about their colormaps :)

**Comment by bryevdv** [comment]
@itsgifnotjiff That's fine, but this is the wrong venue for support with hvplot and holoviews. hvplot is an entirely different project, maintained by an entirely different group of people.
